api/src/apps/host_meta_cache.rs

```diff
@@ -0,0 +1,189 @@
+use crate::apps::Apps;
+use crate::models::service::{Service, ServiceBuilder};
+use crate::models::RequestInfo;
+use crate::models::WebHostMeta;
+
+use chrono::Utc;
+use evmap::ReadHandle;
+use evmap::WriteHandle;
+use multimap::MultiMap;
+use std::convert::From;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::runtime::Runtime;
+use tokio::sync::mpsc;
+use url::Url;
+use yansi::Paint;
+
+pub struct HostMetaCache {
+    reader: ReadHandle<(String, String), Arc<WebHostMeta>>,
+}
+pub struct HostMetaCrawler {
+    writer: WriteHandle<(String, String), Arc<WebHostMeta>>,
+}
+
+pub fn new() -> (HostMetaCache, HostMetaCrawler) {
+    let (reader, writer) = evmap::new();
+
+    (HostMetaCache { reader }, HostMetaCrawler { writer })
+}
+
+impl HostMetaCache {
+    pub fn update_meta_data(
+        &self,
+        services: MultiMap<String, Service>,
+        request_info: &RequestInfo,
+    ) -> MultiMap<String, Service> {
+        let mut assigned_apps = MultiMap::new();
+
+        for (app_name, service) in services.iter_all() {
+            for service in service.iter().cloned() {
+                let tuple = (app_name.clone(), service.id().clone());
+                let mut b = ServiceBuilder::from(service);
+                if let Some(web_host_meta) = self.reader.get_one(&tuple) {
+                    b = b.web_host_meta(web_host_meta.with_base_url(request_info.get_base_url()));
+                }
+                assigned_apps.insert(app_name.clone(), b.build().unwrap());
+            }
+        }
+        assigned_apps
+    }
+}
+
+impl HostMetaCrawler {
+    pub fn spawn(mut self, apps: Arc<Apps>) {
+        let mut runtime = Runtime::new().expect("Should create runtime");
+
+        std::thread::spawn(move || loop {
+            let services = match apps.get_apps() {
+                Ok(apps) => {
+                    info!("apps: {:?}", apps);
+                    apps.iter_all()
+                        .flat_map(|(app_name, services)| {
+                            services
+                                .iter()
+                                // avoid cloning when https://github.com/havarnov/multimap/issues/24 has been implemented
+                                .map(move |service| (app_name.clone(), service.clone()))
+                        })
+                        .collect::<Vec<_>>()
+                }
+                Err(error) => {
+                    error!("error: {}", error);
+                    continue;
+                }
+            };
+
+            let resolved_host_meta_infos = runtime.block_on(Self::resolve_host_meta(services));
+
+            info!("resolved infos: {:?}", resolved_host_meta_infos);
+            self.writer.purge();
```

Comment:
Instead of purging the whole it would be nice to keep existing/valid cache data. This produces a REST request every 5 seconds.